### ISSUE TITLE
Verify Rancher Webhook Rotation workaround and Certificate Renewal

### ISSUE THREAD

**Issue by samjustus** [rancher/rancher#35860]
**A)** Existing workaround: https://rancher.com/docs/rancher/v2.6/en/troubleshooting/expired-webhook-certificates/  does not seem to work when running rancher-webhook:v0.2.0 or v0.2.1. Please validate if the documented workaround works for 2.6.3 and 2.5.11

**Troubleshooting steps:**

Probably able to reproduce as follows:

Provision a Rancher v2.5.x instance and custom cluster
Upgrade to v2.6.x
Check rancher-webhook chart upgraded to 1.0.0+up0.2.0 or 1.0.1+up0.2.1
Move system clock forward over a year
Attempt documented workaround to rotate cattle-webhook-tls
Repro steps: N/A

**B)** If it doesn't work - identify root cause, and and see if below workaround does the trick for both 2.6 and 2.5. 

**Workaround:**
Is workaround available and implemented? Yes

Delete the expired certificate `kubectl delete secret -n cattle-system cattle-webhook-tls`
Get the rancher-webhook deployment and note the current image tag (most likely 0.2.1 or 0.2.0) `kubectl -n cattle-system get deployment rancher-webhook -o yaml`
Edit the rancher-webhook deployment and set the image tag to 0.1.1 `kubectl -n cattle-system edit deployment rancher-webhook`
Once the new rancher-webhook Pod is created and running, validate a new cattle-webhook-tls secret has been created `kubectl -n cattle-system get secret cattle-webhook-tls`
- Edit the rancher-webhook deployment and set the image tag back to the original value `kubectl -n cattle-system edit deployment rancher-webhook`

**C)** Customers installing v2.6.3 (or v2.6.4? - please clarify) will get a certificate that expires in 10 years. However, this won’t help customers that upgrade? Or will upgrade auto-renew cert for 10 years?

**Comment by samjustus** [comment]
Related: 
https://github.com/rancher/rancher/issues/35366
https://github.com/rancher/rancher/issues/35068

**Comment by samjustus** [comment]
Also Max - please look into this `

> "about the webhook - what do 2.5x users do? Do they have to just do the workaround every year or does this partial fix (https://github.com/rancher/rancher/issues/35366) actually have a backport back to 2.5x?"

`

**Comment by genexpr** [comment]
Reproduced workaround (A). Created a 2.5.11 cluster in Docker, made a downstream cluster. Then upgraded Rancher to 2.6.2. Moved the node time one year forward. Creation of a new user failed due to expired web hook certificate.

Followed the steps in the docs of workaround (A) – deleted the cert secret, mutating webhook config, webhook server pod. The workaround works – the three resources got recreated, and I was able to create a new user.

**Comment by genexpr** [comment]
About 2.6.2. I move time one year forward. Then try to create a user and get an expected error (the webhook cert had expired). Then I perform workaround A. Webhook resources are recreated, and that failed user request is automatically picked up, sent again, and succeeds. All is good but the user is shown to be created a year ago. I wonder if this is a problem.

**Comment by genexpr** [comment]
> Also Max - please look into this `
> 
> > "about the webhook - what do 2.5x users do? Do they have to just do the workaround every year or does this partial fix (#35366) actually have a backport back to 2.5x?"
> 
> `

In 2.5.11, Workaround A works and makes the cert valid for 1 year.

**Comment by genexpr** [comment]
Tested different versions of HA Rancher without upgrades, just moving time forward one year on all Control Plane nodes.
In summary, the workaround must trigger the recreation of a webhook pod, which in turn triggers the re-creation of `cattle-webhook-tls` secret containing the cert. The main problem is that the pod does not always get created, even though its ReplicaSet controller does not produce errors, showing that the number of desired replicas is 0. Suddenly, after a delay, it has an event saying it's not meeting its spec, after which the pod gets created and then the cert secret.

### Detailed findings
#### 2.6.0 and 2.6.1
`cattle-webhook-tls` is not immediately re-created. Then on RBAC action:
`Internal error occurred: failed calling webhook "rancherauth.cattle.io": Post "https://rancher-webhook.cattle-system.svc:443/v1/webhook/validation?timeout=10s": context deadline exceeded`
Because the webhook-pod had not been created, as explained above.

In one test, after about 10 minutes, `cattle-webhook-tls` was created (valid for 1 year).
In another test, after about 45 minutes, `cattle-webhook-tls` was created (valid for 1 year).
Once it's created (whenever that happens) then any runs of the workaround trigger the creation of the `cattle-webhook-tls` secret immediately (valid for 1 year).
In a single-node test, `cattle-webhook-tls` was created immediately (valid for 1 year).


##### 2.6.2
`cattle-webhook-tls` is not immediately re-created
After about 45 minutes, `cattle-webhook-tls` was created (valid for 1 year).

##### 2.6.3
`cattle-webhook-tls` is valid for 10 years by default for newly created clusters and instances of the `rancher-webhook` app.

**Comment by samjustus** [comment]
also related: https://jira.suse.com/browse/SURE-3528

**Comment by genexpr** [comment]
More tests show that the workaround works, but with a delay. In a few tests, the webhook pod came up and created the cert secret about 45 minutes after the workaround. Afterwards, any subsequent runs of the workaround work immediately.

**Comment by genexpr** [comment]
The pods being blocked from creation is likely a consequence of moving the time manually forward on Rancher nodes. This is not related to the webhook at all. Disregard the previous comment.

I used 2.6.3, where one can set `CATTLE_NEW_SIGNED_CERT_EXPIRATION_DAYS` to zero for the webhook deployment and immediately invalidate the webhook cert, without moving the time at all. The workaround then works immediately.

I assume the same would be true for 2.6.0-2.6.2, which uses a slightly older version of webhook 0.2.1, but I can’t prove it. Because I would literally have to wait for a year to see if my assumption is correct. 2.6.3 uses a newer version of webhook 0.2.2, which allows to specify `CATTLE_NEW_SIGNED_CERT_EXPIRATION_DAYS`. Previous versions don’t have that, so I can’t be 100% certain if after a year the workaround will work immediately, but I believe it will.

I also believe that the reason people have issues with the pod being up but failing to create a new secret is they miss a step in the docs for the workaround. This is evident from the comments here: https://github.com/rancher/rancher/issues/35068
One might delete the cert secret and the pod, but forget to delete the `mutatingwebhookconfigurations.admissionregistration.k8s.io rancher.cattle.io`.
This comment reminds about it: https://github.com/rancher/rancher/issues/35068#issuecomment-937867815

Although some people say the workaround still works without deleting the configuration, every time I don't delete it, the pod is stuck in trying to create the cert unsuccessfully (2.6.3). And the documented workaround is explicit about deleting all the three things. It was updated only two weeks ago in this PR: https://github.com/rancher/docs/pull/3712

The bottom line: I believe moving the time forward (to simulate cert expiration) causes unrelated issues we don't entirely understand. Without it, the workaround works for 2.5.11 and 2.6.3, but I can't prove it with complete certainty for 2.6.0-2.6.2, where I am forced to move the time forward. But the updated docs PR got confirmations from people that the workaround steps work.

**Comment by brudnak** [comment]
# My checks PASSED

<hr>
  
## Validation Environment

| Component               | Version / Type                          |
| ----------------------- | --------------------------------------- |
| Rancher version         | v2.6.3                                  |
| Rancher cluster type    | Docker                                  |
| Docker version          | 20.10.7, build 20.10.7-0ubuntu5~20.04.2 |
| Helm version            | v2.16.8-rancher1                        |
| Downstream cluster type | N/A                                     |
| Downstream K8s version  | N/A                                     |
| Browser type            | Google Chrome                           |
| Browser version         | 96.0.4664.110 (Official Build) (x86_64) |

## Validation steps

1. Check the cert before the test and take note of the last few digits of the cert
   1. Starting from the **default** Rancher homepage `/dashboard/home`
   2. Click `hamburger menu` -> `local` -> `Download KubeConfig`
   3. Run the following against the local cluster using the kubeconfig file and take note of the last section of the certificate

```shell
kubectl get secrets -n cattle-system cattle-webhook-tls -o json | jq -r ".data.\"tls.crt\"" | base64 -d | openssl x509 -enddate
```

2. Set the `CATTLE_NEW_SIGNED_CERT_EXPIRATION_DAYS` to zero
   1. Starting from the **default** Rancher homepage `/dashboard/home`
   2. Click `hamburger menu` -> `local` -> `Workload` -> `Deployments` -> right click row for `rancher-webhook` in the `cattle-system` namespace -> click `Edit Config`
   3. Under `Environment Variables` -> click `Add Variable` -> select type `Key/Value Pair`
      - Variable Name: `CATTLE_NEW_SIGNED_CERT_EXPIRATION_DAYS`
      - Value: `0` 
   4. Click `save` -> redirected back to the deployments page
   5. Now check the local cluster again using the same command and see that it has changed

```shell
kubectl get secrets -n cattle-system cattle-webhook-tls -o json | jq -r ".data.\"tls.crt\"" | base64 -d | openssl x509 -enddate
```

3. Apply the workaround listed at [https://rancher.com/docs/rancher/v2.6/en/troubleshooting/expired-webhook-certificates/](https://rancher.com/docs/rancher/v2.6/en/troubleshooting/expired-webhook-certificates/) by running the following commands

```shell
kubectl delete secret -n cattle-system cattle-webhook-tls
kubectl delete mutatingwebhookconfigurations.admissionregistration.k8s.io --ignore-not-found=true rancher.cattle.io
kubectl delete pod -n cattle-system -l app=rancher-webhook
```
4. After run this command again and verify that the cert has been updated 

```shell
kubectl get secrets -n cattle-system cattle-webhook-tls -o json | jq -r ".data.\"tls.crt\"" | base64 -d | openssl x509 -enddate
```

**Comment by cbron** [comment]
@brudnak can you test on 2.5.11 also. You won't be able to use the ENV var, but just ensuring that the cert updates to a newer one is enough.

**Comment by brudnak** [comment]
## PASSED | Validation Environment for Workaround and Certificate Renewal Checks

| Component               | Version / Type                          |
| ----------------------- | --------------------------------------- |
| Rancher version         | v2.5.11                                 |
| Rancher cluster type    | High Availability                       |
| Helm version            | v2.16.8-rancher1                        |
| Downstream cluster type | N/A                                     |
| Downstream K8s version  | N/A                                     |
| Browser type            | Google Chrome                           |
| Browser version         | 96.0.4664.110 (Official Build) (x86_64) |

## Validation steps

1. From local cluster dashboard, retrieve the `Kubeconfig File`
2. Run the following command and take note of the output

```shell
kubectl get secrets -n cattle-system cattle-webhook-tls -o json | jq -r ".data.\"tls.crt\"" | base64 -d | openssl x509 -enddate
```

3. Follow the workaround listed at [https://rancher.com/docs/rancher/v2.6/en/troubleshooting/expired-webhook-certificates/](https://rancher.com/docs/rancher/v2.6/en/troubleshooting/expired-webhook-certificates/)

```shell
kubectl delete secret -n cattle-system cattle-webhook-tls
kubectl delete mutatingwebhookconfigurations.admissionregistration.k8s.io --ignore-not-found=true rancher.cattle.io
kubectl delete pod -n cattle-system -l app=rancher-webhook
```

4. Run the command again from step `#2` and compare the output.
  
#### RESULTS

The cert has been successfully updated!
